cpp/src/neighbors/detail/vamana/vamana_serialize.cuh

```diff
@@ -36,6 +36,247 @@
 
 namespace cuvs::neighbors::vamana::detail {
 
+/**
+ * Save the dataset to file.
+ *
+ * Experimental, both the API and the serialization format are subject to change.
+ *
+ * @param[in] res the raft resource handle
+ * @param[in] dataset the raw dataset from which the VAMANA index is built, or the quantized dataset
+ * @param[out] dataset_base_file the path for writing the output file
+ *
+ */
+template <typename T>
+void serialize_dataset(raft::resources const& res,
+                       const cuvs::neighbors::dataset<int64_t>* dataset,
+                       const std::string& dataset_base_file)
+{
+  // try allocating a buffer for the dataset on host
+  try {
+    const cuvs::neighbors::strided_dataset<T, int64_t>* strided_dataset =
+      dynamic_cast<const cuvs::neighbors::strided_dataset<T, int64_t>*>(dataset);
+    if (strided_dataset == nullptr) {
+      RAFT_LOG_DEBUG("dynamic_cast to strided_dataset failed");
+    } else {
+      auto h_dataset =
+        raft::make_host_matrix<T, int64_t>(strided_dataset->n_rows(), strided_dataset->dim());
+      raft::copy(h_dataset.data_handle(),
+                 strided_dataset->view().data_handle(),
+                 strided_dataset->n_rows() * strided_dataset->dim(),
+                 raft::resource::get_cuda_stream(res));
+      std::ofstream dataset_of(dataset_base_file, std::ios::out | std::ios::binary);
+      if (!dataset_of) { RAFT_FAIL("Cannot open file %s", dataset_base_file.c_str()); }
+      size_t dataset_file_offset = 0;
+      int size                   = static_cast<int>(strided_dataset->n_rows());
+      int dim                    = static_cast<int>(strided_dataset->dim());
+      dataset_of.seekp(dataset_file_offset, dataset_of.beg);
+      dataset_of.write((char*)&size, sizeof(int));
+      dataset_of.write((char*)&dim, sizeof(int));
+      for (int i = 0; i < size; i++) {
+        dataset_of.write((char*)(h_dataset.data_handle() + i * h_dataset.extent(1)),
+                         dim * sizeof(T));
+      }
+      dataset_of.close();
+      if (!dataset_of) { RAFT_FAIL("Error writing output %s", dataset_base_file.c_str()); }
+    }
+  } catch (std::bad_alloc& e) {
+    RAFT_LOG_INFO("Failed to serialize dataset");
+  } catch (raft::logic_error& e) {
+    RAFT_LOG_INFO("Failed to serialize dataset");
+  }
+}
+
+/**
+ * Save the index to file with sector alignment.
+ *
+ * Experimental, both the API and the serialization format are subject to change.
+ *
+ * @param[in] res the raft resource handle
+ * @param[in] h_graph the host matrix representation of the VAMANA index
+ * @param[in] dataset the raw dataset from which the VAMANA index is built
+ * @param[in] medoid the medoid
+ * @param[out] output_writer the stream object for writing the output file
+ *
+ */
+template <typename T, typename IdxT, typename HostMatT>
+void serialize_sector_aligned(raft::resources const& res,
```

Comment:
The files created, their names, and how to use them with CPU DiskANN search is quite complex. Added an issue to add documentation for this as a future item: https://github.com/rapidsai/cuvs/issues/906
Not a blocker for this PR, though.